Implement From<std::ops::Range> for Range

vks opened the pull request:
This allows the following syntax:

```rust
let between = Range::from(10..10000);
```

pitdicker commented on src/distributions/range.rs:
```diff
@@ -314,6 +314,19 @@ macro_rules! range_int_impl {
             }
         }
+
+        impl From<::core::ops::Range<$ty>> for RangeInt<$ty> {
```
Can you also implement `From<::core::ops::Range<$ty>>` for `RangeFloat`?

vks replied:
I'm not sure this is a good idea, since we don't have an implementation for closed ranges. On the other hand, this also affects `Range::new` and `Range::new_inclusive`, which are currently identical for floats.

dhardy replied:
And here it's annoying that the `a..b` syntax is half-open — if it were closed it wouldn't be confusing using the syntax for sampling both float and int ranges. As it is though I don't know.

pitdicker commented on src/distributions/range.rs:
```diff
@@ -314,6 +314,19 @@ macro_rules! range_int_impl {
+            }
+        }
+
+        impl From<::core::ops::Range<$ty>> for Range<$ty>
```
Maybe make this a generic implementation for all `RangImpl`s?
```rust
impl<X> From<::core::ops::Range<X>> for Range<X>
    where X: SampleRange,
          <X as SampleRange>::Impl: From<::core::ops::Range<X>>
{
    fn from(r: ::core::ops::Range<X>) -> Range<X> {
        Range { inner: <X as SampleRange>::Impl::from(r) }
    }
}
```

pitdicker replied:
Thinking about it a bit more, a simple generic implementation is all that is needed, no implementation on `RangeImpl` is necessary:
```rust
impl<X: SampleRange> From<::core::ops::Range<X>> for Range<X> {
    fn from(r: ::core::ops::Range<X>) -> Range<X> {
        Range::new(r.start, r.end)
    }
}
```

vks replied:
Sure!

pitdicker commented on src/distributions/range.rs:
```diff
@@ -454,9 +469,13 @@ macro_rules! range_float_impl {
-                // Doing multiply before addition allows some architectures to
-                // use a single instruction.
-                value1_2 * self.scale + self.offset
+                value1_2.mul_add(self.scale, self.offset)
```
`mul_add` is much slower than a separate multiply and addition, because it tries to minimize the rounding error. LLVM can turn the two instructions into one for the platforms that support it.

vks replied:
> mul_add is much slower

Not according to the [documentation](https://doc.rust-lang.org/nightly/std/primitive.f64.html#method.mul_add).

pitdicker replied:
Yes, I read that and tried it before. Can you benchmark it otherwise, because I believe that documentation is wrong.

vks replied:
Actually, when using `target-cpu=native`, `mul_add` is faster for `f32`:
```
test distr_range_f32            ... bench:       1,163 ns/iter (+/- 17) = 3439 MB/s
test distr_range_f64            ... bench:       3,852 ns/iter (+/- 6) = 2076 MB/s
```

vks commented:
> Would it even be possible implement the Standard distribution for core::ops::Range, so that you can write rng.gen(0..10)?

How would that work? `Standard` is not parametrized and `Distribution<Range<f64>>` can only generate `Range<f64>`.

pitdicker commented:
I edited the comment, I'm afraid I'll never learn...

vks commented:
I removed `mul_add` and updated the comment. I think we should not use it, because it is not available without `std` and we can't fall back to `a * b + c` because it is not equivalent.

pitdicker commented:
What did you think of my comment to not implement range for the `RangeImpl`s, but only for `Range`? That would also make it easily available for user types.

vks commented:
Sorry, I missed that comment. I like it!

pitdicker commented on src/distributions/range.rs:
```diff
@@ -454,11 +466,18 @@ macro_rules! range_float_impl {
                 let value1_2 = (rng.$next_u() >> $bits_to_discard)
                                .into_float_with_exponent(0);
-                // Doing multiply before addition allows some architectures to
-                // use a single instruction.
```
Before https://github.com/rust-lang-nursery/rand/search?q=port+range&type=Issues we would do `r.low + r.range * rng.gen::<$ty>()`. I think it is nice to preserve this comment that explains why we choose a different order.

vks replied:
Could you please be more specific? I'm not sure the previous comment was correct. Without `fast-math` it is not correct if the compiler uses FMA.

pitdicker replied:
Ah, there is an instruction that does a true FMA, with the benefit of extra precision, and a simple instruction that produuces exactly the same results as multiply and addition. At least on ARM.

vks replied:
Thanks, I'll add a comment!

pitdicker commented on src/distributions/range.rs:
```diff
@@ -454,11 +466,18 @@ macro_rules! range_float_impl {
             }
         }
+
+        impl From<::core::ops::Range<$ty>> for RangeFloat<$ty> {
```
I don't see a point in keeping these when there also is the more generic implementation?

vks replied:
Removing this would break `RangeFloat::from(range)`, wouldn't it? I would rather keep it, if `RangeFloat` stays public.

pitdicker replied:
`RangeFloat` is supposed to be an implementation detail, only there to make it easier to implement `Range` for user types. No one is supposed to use it directly.

vks replied:
Then it should be private, right? If we make it private (or hide the docs) I would agree that the other impls should be removed.
